### DISCUSSION TITLE
Building wheel for hnswlib (pyproject.toml) ... error

### DISCUSSION THREAD

**Discussion by xkilumax** [imartinez/privateGPT#583] [Q&A]
When I run pip install -r requirements everything runs good until it goes to build the wheel for hnswlib. 

I get the following output:
`Building wheels for collected packages: llama-cpp-python, hnswlib
  Building wheel for llama-cpp-python (pyproject.toml) ... done
  Created wheel for llama-cpp-python: filename=llama_cpp_python-0.1.50-cp311-cp311-win32.whl size=249425 sha256=bc20e8582069d52808843eb5a3775056101af8304c1a1cedb7106e8894c15418
  Stored in directory: c:\users\bgentry\appdata\local\packages\pythonsoftwarefoundation.python.3.11_qbz5n2kfra8p0\localcache\local\pip\cache\wheels\b1\5d\b9\233531d6ca06624db502af0ca6523faa00b6d4d5cec8ab09eb
  Building wheel for hnswlib (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for hnswlib (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [233 lines of output]
      running bdist_wheel
      running build
      running build_ext
      building 'hnswlib' extension
      creating build
      creating build\temp.win32-cpython-311
      creating build\temp.win32-cpython-311\Release
      creating build\temp.win32-cpython-311\Release\python_bindings
      "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\bin\HostX86\x86\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -IC:\Users\bgentry\AppData\Local\Temp\pip-build-env-e3mu0kt1\overlay\Lib\site-packages\pybind11\include -IC:\Users\bgentry\AppData\Local\Temp\pip-build-env-e3mu0kt1\overlay\Lib\site-packages\numpy\core\include -I./hnswlib/ "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0\include" "-IC:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0\Include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\ATLMFC\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\include\um" /EHsc /Tp./python_bindings/bindings.cpp /Fobuild\temp.win32-cpython-311\Release\./python_bindings/bindings.obj /EHsc /openmp /O2 /DVERSION_INFO=\\\"0.7.0\\\"
      bindings.cpp
      ./python_bindings/bindings.cpp(876): warning C4996: 'pybind11_init': PYBIND11_PLUGIN is deprecated, use PYBIND11_MODULE
      ./python_bindings/bindings.cpp(877): warning C4996: 'pybind11::module_::module_': Use PYBIND11_MODULE or module_::create_extension_module instead
      ./python_bindings/bindings.cpp(619): warning C4018: '<=': signed/unsigned mismatch
      ./python_bindings/bindings.cpp(600): note: while compiling class template member function 'pybind11::object Index<float,float>::knnQuery_return_numpy(pybind11::object,size_t,int,const std::function<bool (hnswlib::labeltype)> &)'
      ./python_bindings/bindings.cpp(953): note: see the first reference to 'Index<float,float>::knnQuery_return_numpy' in 'pybind11_init'
      C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\include\type_traits(544): note: see reference to class template instantiation 'Index<float,float>' being compiled
      C:\Users\bgentry\AppData\Local\Temp\pip-build-env-e3mu0kt1\overlay\Lib\site-packages\pybind11\include\pybind11\pybind11.h(1509): note: see reference to class template instantiation 'std::is_polymorphic<Index<float,float>>' being compiled
      ./python_bindings/bindings.cpp(879): note: see reference to class template instantiation 'pybind11::class_<Index<float,float>>' being compiled
      ./python_bindings/bindings.cpp(261): warning C4018: '<=': signed/unsigned mismatch
      ./python_bindings/bindings.cpp(248): note: while compiling class template member function 'void Index<float,float>::addItems(pybind11::object,pybind11::object,int,bool)'
      ./python_bindings/bindings.cpp(953): note: see the first reference to 'Index<float,float>::addItems' in 'pybind11_init'
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\bruteforce.h(105): warning C4018: '<': signed/unsigned mismatch
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\bruteforce.h(101): note: while compiling class template member function 'std::priority_queue<std::pair<dist_t,hnswlib::labeltype>,std::vector<std::pair<dist_t,hnswlib::labeltype>,std::allocator<std::pair<dist_t,hnswlib::labeltype>>>,std::less<_Ty>> hnswlib::BruteforceSearch<dist_t>::searchKnn(const void *,size_t,hnswlib::BaseFilterFunctor *) const'
              with
              [
                  dist_t=float,
                  _Ty=std::pair<float,hnswlib::labeltype>
              ]
      ./python_bindings/bindings.cpp(841): note: see the first reference to 'hnswlib::BruteforceSearch<dist_t>::searchKnn' in 'BFIndex<float,float>::knnQuery_return_numpy'
              with
              [
                  dist_t=float
              ]
      ./python_bindings/bindings.cpp(965): note: see the first reference to 'BFIndex<float,float>::knnQuery_return_numpy' in 'pybind11_init'
      ./python_bindings/bindings.cpp(841): note: see reference to class template instantiation 'hnswlib::BruteforceSearch<dist_t>' being compiled
              with
              [
                  dist_t=float
              ]
      ./python_bindings/bindings.cpp(820): note: while compiling class template member function 'pybind11::object BFIndex<float,float>::knnQuery_return_numpy(pybind11::object,size_t,const std::function<bool (hnswlib::labeltype)> &)'
      ./python_bindings/bindings.cpp(965): note: see the first reference to 'BFIndex<float,float>::knnQuery_return_numpy' in 'pybind11_init'
      C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\include\type_traits(544): note: see reference to class template instantiation 'BFIndex<float,float>' being compiled
      C:\Users\bgentry\AppData\Local\Temp\pip-build-env-e3mu0kt1\overlay\Lib\site-packages\pybind11\include\pybind11\pybind11.h(1509): note: see reference to class template instantiation 'std::is_polymorphic<BFIndex<float,float>>' being compiled
      ./python_bindings/bindings.cpp(957): note: see reference to class template instantiation 'pybind11::class_<BFIndex<float,float>>' being compiled
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\bruteforce.h(113): warning C4018: '<': signed/unsigned mismatch
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(973): warning C4267: 'argument': conversion from 'size_t' to 'unsigned short', possible loss of data
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(908): note: while compiling class template member function 'void hnswlib::HierarchicalNSW<dist_t>::updatePoint(const void *,hnswlib::tableint,float)'
              with
              [
                  dist_t=float
              ]
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(903): note: see the first reference to 'hnswlib::HierarchicalNSW<dist_t>::updatePoint' in 'hnswlib::HierarchicalNSW<dist_t>::addPoint'
              with
              [
                  dist_t=float
              ]
      ./python_bindings/bindings.cpp(879): note: see reference to class template instantiation 'hnswlib::HierarchicalNSW<dist_t>' being compiled
              with
              [
                  dist_t=float
              ]
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(481): warning C4267: 'argument': conversion from 'size_t' to 'unsigned short', possible loss of data
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(445): note: while compiling class template member function 'hnswlib::tableint hnswlib::HierarchicalNSW<dist_t>::mutuallyConnectNewElement(const void *,hnswlib::tableint,std::priority_queue<std::pair<dist_t,hnswlib::labeltype>,std::vector<std::pair<dist_t,hnswlib::labeltype>,std::allocator<std::pair<dist_t,hnswlib::labeltype>>>,hnswlib::HierarchicalNSW<dist_t>::CompareByFirst> &,int,bool)'
              with
              [
                  dist_t=float
              ]
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(1166): note: see the first reference to 'hnswlib::HierarchicalNSW<dist_t>::mutuallyConnectNewElement' in 'hnswlib::HierarchicalNSW<dist_t>::addPoint'
              with
              [
                  dist_t=float
              ]
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(875): note: see the first reference to 'hnswlib::HierarchicalNSW<dist_t>::addPoint' in 'hnswlib::HierarchicalNSW<dist_t>::addPoint'
              with
              [
                  dist_t=float
              ]
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(527): warning C4267: 'argument': conversion from 'size_t' to 'unsigned short', possible loss of data
      C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\hnswlib\hnswalg.h(755): warning C4018: '<': signed/unsigned mismatch
      ./python_bindings/bindings.cpp(323): note: see reference to function template instantiation 'std::vector<data_t,std::allocator<data_t>> hnswlib::HierarchicalNSW<dist_t>::getDataByLabel<data_t>(hnswlib::labeltype) const' being compiled
              with
              [
                  data_t=float,
                  dist_t=float
              ]
      ./python_bindings/bindings.cpp(304): note: while compiling class template member function 'std::vector<std::vector<data_t,std::allocator<data_t>>,std::allocator<std::vector<data_t,std::allocator<data_t>>>> Index<float,data_t>::getDataReturnList(pybind11::object)'
              with
              [
                  data_t=float
              ]
      ./python_bindings/bindings.cpp(953): note: see the first reference to 'Index<float,float>::getDataReturnList' in 'pybind11_init'
      creating C:\Users\bgentry\AppData\Local\Temp\pip-install-3bg3q33u\hnswlib_3a3b395aec6549cd9c652365df136e65\build\lib.win32-cpython-311
      "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\bin\HostX86\x86\link.exe" /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO "/LIBPATH:C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0\libs" "/LIBPATH:C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0" "/LIBPATH:C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.11_3.11.1008.0_x64__qbz5n2kfra8p0\PCbuild\win32" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\lib\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22621.0\ucrt\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\\lib\10.0.22621.0\\um\x86" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.36.32532\lib\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.8\lib\um\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.22621.0\ucrt\x86" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\\lib\10.0.22621.0\\um\x86" /EXPORT:PyInit_hnswlib build\temp.win32-cpython-311\Release\./python_bindings/bindings.obj /OUT:build\lib.win32-cpython-311\hnswlib.cp311-win_amd64.pyd /IMPLIB:build\temp.win32-cpython-311\Release\./python_bindings\hnswlib.cp311-win_amd64.lib
         Creating library build\temp.win32-cpython-311\Release\./python_bindings\hnswlib.cp311-win_amd64.lib and object build\temp.win32-cpython-311\Release\./python_bindings\hnswlib.cp311-win_amd64.exp
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThread_tss_create
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_AsEncodedString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyBaseObject_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyMem_Calloc
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_BufferError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_SetAttr
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCFunction_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyWeakref_NewRef
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThreadState_Get
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_AsUnsignedLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_Malloc
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyImport_ImportModule
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyBytes_AsString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_NormalizeException
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_Occurred
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromSsize_t
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_DecodeUTF8
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_AsUTF8AndSize
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyEval_SaveThread
      bindings.obj : error LNK2001: unresolved external symbol __imp___PyThreadState_UncheckedGet
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_RuntimeError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_SetAttrString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GC_UnTrack
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyInstanceMethod_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_MemoryError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_GetItemWithError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromUnsignedLongLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__Py_GetVersion
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_GetContext
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_Contains
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFrame_GetLineNumber
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFloat_AsDouble
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_AsUnsignedLongLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyGILState_Release
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyBytes_AsStringAndSize
      bindings.obj : error LNK2001: unresolved external symbol __imp___Py_NoneStruct
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_AsLongLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyTuple_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GenericSetDict
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_Size
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromLongLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFloat_FromDouble
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyList_Size
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyByteArray_AsString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_SetString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_WriteUnraisable
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_ValueError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromUnsignedLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_Next
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_Format
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThreadState_New
      bindings.obj : error LNK2001: unresolved external symbol __imp___Py_FalseStruct
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFloat_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyMethod_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyType_IsSubtype
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_Restore
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_OverflowError
      bindings.obj : error LNK2001: unresolved external symbol __imp___Py_Dealloc
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyTuple_GetItem
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_GetPointer
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_Fetch
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_CallFunctionObjArgs
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_AsLong
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GenericGetDict
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_ClearWeakRefs
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_FromFormat
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GetBuffer
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyList_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_DecodeUTF16
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyModule_Create2
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyType_Ready
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GetAttrString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyErr_Clear
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyTuple_Size
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThread_tss_get
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyException_SetTraceback
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_GetName
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyBytes_Size
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyList_GetItem
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyException_SetContext
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyInstanceMethod_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_IsInstance
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_DecodeUTF32
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyMem_Free
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyTuple_SetItem
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCMethod_New
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThread_tss_set
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_GetItem
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_CallObject
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_HasAttrString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyProperty_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyIndex_Check
      bindings.obj : error LNK2001: unresolved external symbol __imp___Py_NotImplementedStruct
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyByteArray_Size
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyGILState_Ensure
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_SetContext
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyNumber_Check
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyNumber_Long
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThread_tss_alloc
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThreadState_DeleteCurrent
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_TypeError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyThreadState_Clear
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyEval_GetBuiltins
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_Copy
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_Str
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_AsUTF8String
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyModule_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCallable_Check
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFrame_GetBack
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyFrame_GetCode
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_IndexError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_ImportError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_SetPointer
      bindings.obj : error LNK2001: unresolved external symbol __imp___Py_TrueStruct
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyExc_SystemError
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_SetItem
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyException_SetCause
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyDict_DelItemString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyUnicode_FromString
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyLong_FromSize_t
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyEval_AcquireThread
      bindings.obj : error LNK2001: unresolved external symbol __imp___PyType_Lookup
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyGILState_GetThisThreadState
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyBuffer_Release
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyObject_Repr
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyByteArray_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyNumber_Float
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyType_Type
      bindings.obj : error LNK2001: unresolved external symbol __imp__PySequence_Tuple
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyCapsule_SetName
      bindings.obj : error LNK2001: unresolved external symbol __imp__PyEval_RestoreThread
      bindings.obj : error LNK2001: unresolved external symbol __imp___PyObject_GetDictPtr
      build\lib.win32-cpython-311\hnswlib.cp311-win_amd64.pyd : fatal error LNK1120: 137 unresolved externals
      error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.36.32532\\bin\\HostX86\\x86\\link.exe' failed with exit code 1120
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for hnswlib
Successfully built llama-cpp-python
Failed to build hnswlib
ERROR: Could not build wheels for hnswlib, which is required to install pyproject.toml-based projects`

I've gone through this line by line but I can't find the source of the problem.

**Comment by Vinaykumar-Puppala** [comment]
https://stackoverflow.com/questions/73969269/error-could-not-build-wheels-for-hnswlib-which-is-required-to-install-pyprojec
I tried above suggestion and it worked.  Give a try @xkilumax

Perform below command before pip install :
`export HNSWLIB_NO_NATIVE=1  `

**Comment by xkilumax** [comment]
I have already done all that. I already saw that post and followed each suggestion. My VS has all the needed components. My environment variables have been updated. And the HNSWLIB_NO_NATIVE=1 doesn't work. In fact, it gives the following output:

PS C:\Users\bgentry\Desktop\New_folder\PrivateGPT> export HNSWLIB_NO_NATIVE=1
export : The term 'export' is not recognized as the name of a cmdlet, function, script file, or operable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ export HNSWLIB_NO_NATIVE=1
+ ~~~~~~
    + CategoryInfo          : ObjectNotFound: (export:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
    
    I thought it might be a Python command, so I tried the following:
    
    PS C:\Users\bgentry\Desktop\New_folder\PrivateGPT> python export HNSWLIB_NO_NATIVE=1
C:\Users\bgentry\AppData\Local\Microsoft\WindowsApps\python.exe: can't open file 'C:\\Users\\bgentry\\Desktop\\New_folder\\PrivateGPT\\export': [Errno 2] No such file or directory

**Comment by xkilumax** [comment]
And because I'm using PowerShell it should look like the following:

PS C:\Users\bgentry\Desktop\New_folder\PrivateGPT> $env:VARIABLE_NAME = "HNSWLIB_NO_NATIVE=1"
PS C:\Users\bgentry\Desktop\New_folder\PrivateGPT> pip install -r requirements.txt

But that doesn't fix the problem.

**Comment by pseudotensor** [comment]
I'm curious when  you guys hit this problem.  For h2oGPT I didn't hit any issues with the privateGPT like deps or any others.  Maybe try the same setup described there for privateGPT: https://github.com/h2oai/h2ogpt#windows-1011
